### PR TITLE
Finished Grammars HW

### PR DESCRIPTION
**Pull request by pattisonb**
finished all the questions

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -53,7 +53,8 @@ In one sentence, explain why.
 
 ## A1.2
 
-«replace this with your answer»
+0(1)
+You can compare each character in the first half to its mirror in the second half to check equality and this way you would only ever have two tiles in memory at the same time.
```
Comment (pragdave) [inline README.md:57]: You could if you had everything available. But that's not a parser. 

The parser takes its input one token at a time. It doesn't know where the end is
until it gets there. So the mechanical parsing process is to get the next token,
push it on a stack, and then compare the rest of the input. As a result, the
stack can grow to at least n/2, so the memory consumption will be O(n)

# -3

File: README.md [hunk 2]
```diff
@@ -82,7 +83,15 @@ Write the BNF (not EBNF) description for this language.
...
-«replace this with your answer»
+\<sentence> ::= "the" + \<adjective> + \<noun> +  \<verb> + \<adverb>
+
+\<adjective> ::= \<<empty>> | "lazy" | "smelly"
+
+\<noun> ::= "dog" | "cat"
+
+\<verb> ::= "ate" | "ran"
+
+\<adverb> ::= \<<empty>> | "slowly" | "noisily"
```
Comment (pragdave) [inline README.md:94]: this is almost correct; it just doesn't allow for multiple adjectives.

# -2

File: README.md [hunk 3]
```diff
@@ -109,7 +126,7 @@ Write this grammar using EBNF with common extensions
 
 ## A2.3
 
-«replace this with your answer»
+FSM_Diagram.png
```
Comment (pragdave) [inline README.md:129]: The transition from S3 to S4 on the empty string isn't correct: there can't be a transition that isn't triggered by something. Instead S4 and S3 both have to deal with EOI:

image

# -2

File: README.md [hunk 4]
```diff
@@ -125,7 +142,20 @@ Current state | Next word | Next state
...
+    S0        |    the        |     S1
+    S1        |   lazy	      |   	S1
+    S1        |   smelly      |   	S1
+    S1        |   dog         |     S2
+    S1        |   cat         |     S2
+    S2        |   ate         |    	S3 
+    S2        |   ran         |     S3
+    S3        |    noisily    |     S4
+    S3        |    slowly     |     S4
+    S3        |   <<empty>>   |   	S4
```
Comment (pragdave) [inline README.md:156]: Same comment here, but no deduction because the table reflects the diagram.

File: README.md [hunk 5]
```diff
@@ -145,7 +175,9 @@ code, include a script or makefile that will do the job.
 
 ## A2.5
 
-«replace this with your answer»
+The file is called sentences.cpp.
```
Comment (pragdave) [inline README.md:178]: I appreciate you adding code to test both positive and negative cases. However, these are actually tests, as what they do is write 1s and 0s to the console. A test will check that the code returns what is expected, and report when it doesn't.

Here's what I mean, with a fairly minimal hack to your code.

~~~ c++
using namespace std;
typedef vector<string> sentence;
typedef vector<sentence> sentences;
// ...
int main() {
	sentences valid {
	  {"the", "dog", "ate", "EOI"},
	  {"the", "lazy", "smelly", "dog", "ran", "slowly", "EOI"},
	  {"the", "lazy", "smelly", "lazy", "dog", "ate", "noisily", "EOI"},
	  {"the", "dog", "ate", "slowly", "EOI"},
	  {"the", "dog", "ran", "noisily", "EOI"},
	};
	 sentences invalid {
	  {"dog", "ate", "EOI"},
	  {"the", "lazy", "smelly", "ran", "EOI"},
	  {"the", "lazy", "smelly", "dog", "ate", "slowly", "noisily", "EOI"},
	  {"the", "dog", "EOI"},
	 };

	for (sentence s: valid) {
		for (const auto &e : s) cout << e << " "; cout << endl;
		assert(isSentence(s));
	}
	for (sentence s: invalid) {
		for (const auto &e : s) cout << e << " "; cout << endl;
		assert(!isSentence(s));
	}

	return 0;
};
~~~

# -2

Now on to the parser.

The idea of the exercise was to "Translate this table into a programming language of your choice"; that is to create an actual data structure in C++ and then traverse it during the parse. The next best alternative would be to have written the state machine totally in code, with (maybe) a function per state, or a case statement. But the code here is just a manually written parser without any mention of states. It's actually a good example of why state machines are helpful. This code has if statements nested 7 level deep. Here's a version using a table:

~~~ c++
typedef int state;
typedef std::map<std::string, state> transitions;
typedef std::map<state, transitions> state_table;

const int END = 99;

state_table states = {
  { 0, {{ "the", 1 }}},
	{ 1, {{ "lazy", 2}, {"smelly", 2}, {"dog", 3}, {"cat", 3}}},
	{ 2, {{ "lazy", 2}, {"smelly", 2}, {"dog", 3}, {"cat", 3}}},
	{ 3, {{ "ate", 4},  {"ran",    4}}},
	{ 4, {{ "slowly", 5}, { "noisily", 5}, {"EOI", END}}},
	{ 5, {{ "EOI", END }}}
};


int getNextState(int currentState, string nextWord) {
	transitions t = states[currentState];
	if (t.count(nextWord))
		return t[nextWord];
	else
		return -1;
}
~~~

# -6

File: README.md [hunk 6]
```diff
@@ -164,4 +196,4 @@ explain why.
 
 ## A2.7
 
-«replace this with your answer»
+Type 2 because the langauge can repeat adjectives making it infite therefore it cannot be Type 3.
```
Comment (pragdave) [inline README.md:199]: But is can be parsed using a state machine, which makes it type 3.

# -3
